### ISSUE TITLE
Avoid streaming SQL Server FOR JSON or FOR XML in chunked results

### ISSUE THREAD

**Issue by AnghelLeonard** [jOOQ/jOOQ#11386]
### Expected behavior
Same behavior.

### Actual behavior
Let's consider the following query:

```Java
var result = 
ctx.select(
  PRODUCTLINE.PRODUCT_LINE, 
  PRODUCTLINE.TEXT_DESCRIPTION,
  select(
    PRODUCT.PRODUCT_NAME, 
    PRODUCT.PRODUCT_VENDOR, 
    PRODUCT.QUANTITY_IN_STOCK)
  .from(PRODUCT)
  .where(PRODUCT.PRODUCT_LINE.eq(PRODUCTLINE.PRODUCT_LINE))
  .orderBy(PRODUCT.QUANTITY_IN_STOCK).limit(15) 
  .forJSON().path().asField("products")
)
.from(PRODUCTLINE)
.orderBy(PRODUCTLINE.PRODUCT_LINE).limit(2) 
.forJSON().path()
.fetch();
```

In SQL Server this query produces a `Result<Record1<JSON>>` having the following coordinates:

- The `Result` size is 3. Generally speaking, has a size of `n` and each `n` has a size of 1.99 KB and a total of 2034 characters (exactly as SSMS act by default - https://stackoverflow.com/questions/54973536/for-json-path-results-in-ssms-truncated-to-2033-characters/54973676).

Here is a sample of `result`:

`result[0]=[{"product_line":"Classic Cars", ... :"Highway 66 Mini C`
Notice that `result[1]` continue from where `result[0]` has left:
`result[1]=lassics","quantity_in_stock":3975}, ...`

Trying to convert this into a `String` via `fetchSingleInto()` obviously throws `TooManyRowsException`. To get a single `String` I can join these chunks.

Now, running the same thing is PostgreSQL or Oracle produces a `Result` of size 1 containing the entire JSON. Applying `fetchSingleInto()` works with no problems.

I hope I managed to explain it :)

### Versions

- jOOQ:3.14.7 (org.jooq.trial-java-8)
- Java:14
- Database (include vendor):SQL Server, Oracle, PostgreSQL
- OS: Windows 10
- JDBC Driver (include name if inofficial driver): Microsoft JDBC for SQL Server

**Comment by AnghelLeonard** [comment]
The solution can be a cast `AS XMLDATA`:

https://docs.microsoft.com/en-us/sql/t-sql/queries/select-for-clause-transact-sql?view=sql-server-2017#for-xml

**Comment by lukaseder** [comment]
Thanks a lot for your detailed report. I'll look into this soon - quite likely a bug.

**Comment by lukaseder** [comment]
Oh wow, I can reproduce it now! Weird "feature". It seems to happen only if the top level `SELECT` has a `FOR JSON` clause. E.g. this:

```sql
select count(*) over() as c, row_number() over(order by (select 1)) as rn
from sys.all_objects
for json auto
```

Produces this undesirable result format:

![image](https://user-images.githubusercontent.com/734593/126345633-8d6ec5d4-0481-4230-92b4-f72e3c6e0dd7.png)

But nesting it:

```sql
select (
  select row_number() over(order by (select 1)) as rn
  from sys.all_objects
  for json auto
)
```

Produces the correct one:

![image](https://user-images.githubusercontent.com/734593/126345746-6a3152b7-4960-4e2c-af65-5158b6545121.png)

I've never noticed this in the integration tests, because:

- Most tests are functional tests, and they don't act on such "large" data sets
- The edge case tests acting on "large" data sets have some other things in them, that already produced the nested version

I don't really see the value of this behaviour in jOOQ's case. Might be useful if the document is streamed, but since we're not streaming it, there's no benefit. I guess a new `Settings` will allow for opting out of the bug fix, in case someone *wants* the current behaviour for some reason (though it really doesn't make sense to have non-well-formed content in `org.jooq.JSON` or `org.jooq.XML`)

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0, 3.15.2 (https://github.com/jOOQ/jOOQ/issues/12232), and 3.14.14 (https://github.com/jOOQ/jOOQ/issues/12233). The workaround is to apply the nesting as scalar subquery manually. I've also documented this workaround on the Stack Overflow question you've linked: https://stackoverflow.com/a/68457271/521799

Thanks again for your report!
